src/net/atlas/download.rs

```diff
@@ -0,0 +1,1025 @@
+        self.retry_count += 1;
+    }
+
+    pub fn has_fully_succeed(&self) -> bool {
+        self.attachments_instances_count == 0
+    }
+
+    pub fn get_missing_pages_for_contract_id(
+        &self,
+        contract_id: &QualifiedContractIdentifier,
+    ) -> Vec<u32> {
+        let mut pages_indexes = HashSet::new();
+        if let Some(missing_attachments) = self.attachments_instances.get(&contract_id) {
+            for ((page_index, _), _) in missing_attachments.iter() {
+                pages_indexes.insert(*page_index);
+            }
+        }
+        pages_indexes.into_iter().collect()
+    }
+
+    pub fn resolve_attachment(&mut self, content_hash: &Hash160) {
+        for missing_attachments in self.attachments_instances.values_mut() {
+            let mut keys = vec![];
+            for (k, hash) in missing_attachments.iter() {
+                if hash == content_hash {
+                    keys.push(k.clone());
+                }
+            }
+            for key in keys {
+                missing_attachments.remove(&key);
+                self.attachments_instances_count =
+                    self.attachments_instances_count.saturating_sub(1);
+            }
+        }
+    }
+
+    pub fn get_stacks_block_id(&self) -> StacksBlockId {
+        StacksBlockHeader::make_index_block_hash(&self.consensus_hash, &self.block_header_hash)
+    }
+}
+
+impl Ord for AttachmentsBatch {
+    fn cmp(&self, other: &AttachmentsBatch) -> Ordering {
+        other
+            .retry_count
+            .cmp(&self.retry_count)
+            .then_with(|| {
+                self.attachments_instances_count
+                    .cmp(&other.attachments_instances_count)
+            })
+            .then_with(|| other.block_height.cmp(&self.block_height))
+    }
+}
+
+impl PartialOrd for AttachmentsBatch {
+    fn partial_cmp(&self, other: &AttachmentsBatch) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ReliabilityReport {
+    pub total_requests_sent: u32,
+    pub total_requests_success: u32,
+}
+
+impl ReliabilityReport {
+    pub fn bump_successful_requests(&mut self) {
+        self.total_requests_sent += 1;
+        self.total_requests_success += 1;
+    }
+
+    pub fn bump_failed_requests(&mut self) {
+        self.total_requests_sent += 1;
+    }
+}
+
+impl ReliabilityReport {
+    pub fn new(total_requests_sent: u32, total_requests_success: u32) -> ReliabilityReport {
+        ReliabilityReport {
+            total_requests_sent,
+            total_requests_success,
+        }
+    }
+
+    pub fn empty() -> ReliabilityReport {
+        ReliabilityReport {
+            total_requests_sent: 0,
+            total_requests_success: 0,
+        }
+    }
+
+    pub fn score(&self) -> u32 {
+        match self.total_requests_sent {
+            0 => 0 as u32,
+            n => (self.total_requests_success * 1000 / n * 1000) + n,
```

Comment:
I think you may be missing some `()`'s here.

Reply:
Good catch, thanks! Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/003b699f2af3fb08c529da793777a4c978cec7e6